### PR TITLE
[DOC] Guidelines for downgrade

### PR DESCRIPTION
**Pull request by PaulRMellor**
Signed-off-by: prmellor <pmellor@redhat.com>

**Documentation**

New section in the _Deploying and Upgrading_ guide to provide guidance on how to revert to previous versions of Operators and Kafka. 

Described in two stages like the upgrade: Cluster Operator, then Kafka downgrade (optionally). The CO procedure reverses the upgrade steps. Downgrade of Kafka mentions reverting `genericKafkaListener` as a step (but no mention of API versions).

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Comment by PaulRMellor** [comment]
When ready to merge, please cherry-pick to 0.20

File: documentation/assemblies/upgrading/assembly-downgrade-kafka-versions.adoc [hunk 1]
```diff
@@ -1,20 +1,13 @@
...
 
 Kafka version downgrades are performed using the Cluster Operator.
 
-Whether and how the Cluster Operator performs a downgrade depends on the differences between versions of:
-
-* Interbroker protocol
-* Log message format
-* ZooKeeper
-
+//Version constraints on the downgrade
```
Comment (scholzj) [inline documentation/assemblies/upgrading/assembly-downgrade-kafka-versions.adoc:10]: Should this still be here?

Comment (PaulRMellor) [reply documentation/assemblies/upgrading/assembly-downgrade-kafka-versions.adoc:10]: Note on the file included in the assembly

File: documentation/assemblies/upgrading/assembly-downgrade-kafka-versions.adoc [hunk 2]
```diff
@@ -1,20 +1,13 @@
...
-// assembly-upgrade.adoc
-
-
-[id='assembly-downgrading-kafka-versions-{context}']
+// assembly-downgrade.adoc
 
+[id='assembly-downgrade-kafka-versions-{context}']
 = Downgrading Kafka
 
 Kafka version downgrades are performed using the Cluster Operator.
```
Comment (laidan6000) [inline documentation/assemblies/upgrading/assembly-downgrade-kafka-versions.adoc:8]: _Suggested change:_

performed by the Cluster Operator.

File: documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc [hunk 3]
```diff
@@ -8,30 +8,11 @@
...
-----
-
-The downgrade would not be possible if the `log.message.format.version` was set at `"{LogMsgVersHigher}"` or a value was absent (so that the parameter took the default value for a {KafkaVersionHigher} broker of {LogMsgVersHigher}).
-====
-
 .Prerequisites
 
 For the `Kafka` resource to be downgraded, check:
 
+* [IMPORTANT] xref:com-target-downgrade-version-{context}[Compatibility of Kafka versions].
```
Comment (laidan6000) [inline documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc:15]: I'm not sure about the square bracket convention here. Perhaps bold is better?

File: documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc [hunk 4]
```diff
@@ -8,30 +8,11 @@
...
 * The Cluster Operator, which supports both versions of Kafka, is up and running.
 * The `Kafka.spec.kafka.config` does not contain options that are not supported in the version of Kafka you are downgrading to.
```
Comment (laidan6000) [inline documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc:17]: _Suggested change:_

Use the same phrase in lines 17 and 18:

Kafka version being downgraded to

File: documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc [hunk 5]
```diff
@@ -52,4 +52,4 @@ You now have an updated Cluster Operator, but the version of Kafka running in th
 
 .What to do next
 
-Following the Cluster Operator upgrade, you can perform a xref:assembly-upgrading-kafka-versions-str[Kafka upgrade].
+Following the Cluster Operator upgrade, you can perform a xref:assembly-upgrading-kafka-versions-str[Kafka downgrade].
```
Comment (laidan6000) [inline documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:55]: I would just remove what to do next from this module.

File: documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+kubectl get po my-cluster-kafka-0 -o jsonpath='{.spec.containers[0].image}'
+----
++
+The image tag shows the new Strimzi version followed by the Kafka version. For example, `_NEW-STRIMZI-VERSION_-kafka-_CURRENT-KAFKA-VERSION_`.
+
+You now have an updated Cluster Operator, but the version of Kafka running in the cluster it manages is unchanged.
+
+.What to do next
+
+Following the Cluster Operator downgrade, you can perform a xref:assembly-downgrade-kafka-versions-{context}[Kafka downgrade].
```
Comment (laidan6000) [inline documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:51]: You can or you must?

**Review by laidan6000 — CHANGES_REQUESTED** [review]
New procedure looks good. Made a few comments.

File: documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc (hunk 3 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc:15]: I tried the bold, but thought just the capital was enough in the end. We use a similar format elsewhere, like with _ADVANCED OPTION:_ for callouts

image

File: documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc (hunk 4 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc:17]: :+1:  I changed both to: _by the Kafka version being downgraded to._

File: documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc (hunk 5 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:55]: Assuming you mean just the title. Removed.

File: documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc (hunk 6 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:51]: Can, as in it's now possible after the CO downgrade.

File: documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc (hunk 5 above)
Comment (laidan6000) [reply documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc]: I actually meant this whole thing! Why would a user downgrade Kafka after they've upgraded the Cluster Operator?

File: documentation/assemblies/upgrading/assembly-downgrade.adoc [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+If you are encountering issues with the latest version of Strimzi,
+you can revert your installation to the previous version.
+
+As with the Strimzi upgrade process, you can perform your downgrade in two stages:
+
+. Revert your Cluster Operator to the previous Strimzi version.
+** xref:proc-downgrade-cluster-operator-{context}[]
+
+. Downgrade all Kafka brokers and client applications to the previous Kafka version.
```
Comment (tombentley) [inline documentation/assemblies/upgrading/assembly-downgrade.adoc:16]: This is not correct. The operator currently takes charge of the interbroker protocol, and updates it automatically. That means that the user cannot safely downgrade the brokers, even if they didn't complete the upgrade. I.e. once they've started to upgrade the Kafka cluster then they can't currently safely back out. They can upgrade and downgrade the operator.

Comment (PaulRMellor) [reply documentation/assemblies/upgrading/assembly-downgrade.adoc:16]: @tombentley Thanks Tom. I thought we could perform the downgrade if interbroker versions were the same. So you recommend we drop the section on downgrading Kafka entirely for now?

Comment (tombentley) [reply documentation/assemblies/upgrading/assembly-downgrade.adoc:16]: @PaulRMellor yeah, it would work like that if the operator didn't bump the interbroker version for you. IOW once you've changed `Kafka.spec.kafka.version` you've essentially changed the interbroker version, and Kafka downgrade is no longer possible. So, yes we should drop the section on downgrading Kafka for the time being.

**Comment by PaulRMellor** [comment]
To reflect the current situation with the interbroker protocol and `Kafka.spec.kafka.version`, 
I've dropped assemblies/upgrading/assembly-downgrade.adoc from the /deploying master.adoc, which keeps the Kafka downgrade content out of the build. I've added the _Downgrading the Cluster Operator to a previous version_ procedure directly to the /deploying master.adoc. We can reintroduce the Kafka downgrade sections when appropriate. I've moved the warning on not being able to revert the Kafka upgrade in the _Upgrading Kafka brokers and client applications_ procedure to the step where change `Kafka.spec.kafka.version`.

**Comment by PaulRMellor** [comment]
@tombentley @ppatierno @scholzj - looking for any more comments on **_proc-downgrade-cluster-operator.adoc_** and
**_proc-upgrade-brokers-newer-kafka.adoc_** (The other files are left out the doc until we can perform a kafka downgrade)

File: documentation/modules/upgrading/con-downgrade-target-version.adoc [hunk 8]
```diff
@@ -4,15 +4,40 @@
...
+Kafka downgrades are dependent on compatible current and target xref:ref-kafka-versions-{context}[Kafka versions],
+and the state at which messages have been logged.
 
-* If the target downgrade version of Kafka has the same `log.message.format.version` as the current version, the Cluster Operator downgrades by performing a single rolling restart of the brokers.
-* If the target downgrade version of Kafka has a different `log.message.format.version`, downgrading is only possible if the running cluster has _always_ had `log.message.format.version` set to the version used by the downgraded version.
-+
+You cannot revert to the previous Kafka version if the `inter.broker.protocol.version` differs, or messages have been added
+to message logs that use a newer `log.message.format.version`.
+
+The `inter.broker.protocol.version` determines the schema used in the messages written to `__consumer_offsets`.
```
Comment (tombentley) [inline documentation/modules/upgrading/con-downgrade-target-version.adoc:15]: It's not _just_ this, it's used to gate other persistent data too (such as the versions of JSON persisted in zookeeper).

File: documentation/modules/upgrading/con-downgrade-target-version.adoc [hunk 9]
```diff
@@ -4,15 +4,40 @@
...
+If you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the broker will not understand the messages.
```
Comment (tombentley) [inline documentation/modules/upgrading/con-downgrade-target-version.adoc:16]: While that's true, it doesn't inform the reader about what the consequences of that will be. We should be clear that, in general, it  basically results in a "broker cluster". The rest of this concept then goes into detailed talk about the log message format version which could easily mislead people into thinking this sentence is not really a problem. I think we should probably omit the following sentences because they're mostly irrelevant due to a broker cluster generally being a "game over" situation.

File: documentation/modules/upgrading/con-downgrade-target-version.adoc (hunk 8 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/con-downgrade-target-version.adoc:15]: How about:

> The `inter.broker.protocol.version` determines the schemas used for persistent metadata stored by the broker, such as the schema for messages written to `__consumer_offsets`.

Unless we want to go into more detail?

Comment (PaulRMellor) [reply documentation/modules/upgrading/con-downgrade-target-version.adoc:15]: This section is being left out for now

File: documentation/modules/upgrading/con-downgrade-target-version.adoc (hunk 9 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/con-downgrade-target-version.adoc:16]: We're leaving this section out for now. I'm not sure how much we will want to retain when this situation is rectified?

Comment (PaulRMellor) [reply documentation/modules/upgrading/con-downgrade-target-version.adoc:16]: I've updated this sentence to:

> Therefore, if you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the cluster will not work.

File: documentation/assemblies/upgrading/assembly-downgrade.adoc [hunk 10]
```diff
@@ -0,0 +1,26 @@
+// This assembly is included in the following assemblies:
+//
+// deploying/master.adoc
+
+[id='assembly-downgrade-{context}']
+= Downgrading Strimzi
+
+If you are encountering issues with the latest version of Strimzi,
```
Comment (scholzj) [inline documentation/assemblies/upgrading/assembly-downgrade.adoc:8]: The `latest` version of Strimzi sounds weird given the documentation is for the `latest` Strimzi version for a month or two but then it is for some older version but might be still use. Can we maybe rephrase it to say something like _If you are encountering issues with the version of Strimzi you upgraded to_?

File: documentation/assemblies/upgrading/assembly-downgrade.adoc [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+you can revert your installation to the previous version.
+
+As with the Strimzi upgrade process, you can perform your downgrade in two stages:
+
+. Revert your Cluster Operator to the previous Strimzi version.
+** xref:proc-downgrade-cluster-operator-{context}[]
+
+. Downgrade all Kafka brokers and client applications to the previous Kafka version.
+** xref:assembly-downgrade-kafka-versions-{context}[]
```
Comment (scholzj) [inline documentation/assemblies/upgrading/assembly-downgrade.adoc:17]: I think this will be confusing ... saying it happens in two stages for me gives it a particular order. But the order is not so simple. For example, I'm running Strimzi 0.19.0 with Kafka 2.5. And I upgrade first to Strimzi 0.20 + Kafka 2.5 and then to Kafka 2.6. In that case, the order here is actually first to downgrade Kafka 2.6 to 2.5 if still possible and then to downgrade Strimzi 0.20 to 0.19.

File: documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc [hunk 12]
```diff
@@ -70,6 +51,11 @@ NOTE: You must format the value of `log.message.format.version` as a string to p
 +
 See xref:con-versions-and-images-str[]
 
+.. If you changed the Kafka listener configuration to use the array format of the `GenericKafkaListener` schema,
+revert the `Kafka.spec.kafka.listeners` to use the old format.
++
+See xref:con-upgrade-listeners-str[]
```
Comment (scholzj) [inline documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc:57]: This has to be part of the Cluster Operator downgrade. It does not have to be changed with Kafka downgrade. The operator supports the new format with any Kafka version it supports.

I also wonder if we should make it more generic: _Revert any changes to the CRs_ or something. As there are more new things in each release. E.g. rack awareness in Connect, host aliases and other things are not supported after the operator downgrade and need to be removed. So making it more generic would be IMHO a plus.

File: documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc [hunk 13]
```diff
@@ -0,0 +1,50 @@
+// Module included in the following assemblies:
+//
+// assembly-downgrade.adoc
+
+[id='proc-downgrade-cluster-operator-{context}']
+= Downgrading the Cluster Operator to a previous version
+
+If you are encountering issues with the latest version of Strimzi,
```
Comment (scholzj) [inline documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:8]: Same as earlier ... the _latest version_ sounds weird and confusing here.

File: documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc [hunk 14]
```diff
@@ -67,6 +68,10 @@ spec:
 ----
 <1> This is changed to the new version
 <2> This remains at the current version
+--
++
+WARNING: You cannot downgrade Kafka once `spec.kafka.version` has changed.
```
Comment (scholzj) [inline documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:73]: Is this because of the `inter.broker.protocol.version`? Or why is it the case?

Comment (tombentley) [reply documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:73]: `inter.broker.protocol.version` indeed

Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:73]: Changed to:

> WARNING: You cannot downgrade Kafka once `spec.kafka.version` has changed, as the `inter.broker.protocol.version` for the new Kafka version will differ.  The protocol version determines the schemas used for persistent metadata stored by the broker, including messages written to `__consumer_offsets`. The cluster will not understand the messages.

Comment (scholzj) [reply documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:73]: Are we cherry-picking this for 0.20 release branch? If not, maybe we should think about this a bit more to improve the experience of this before we do 0.21 and before we merge this.

File: documentation/assemblies/upgrading/assembly-downgrade.adoc (hunk 11 above)
Comment (PaulRMellor) [reply documentation/assemblies/upgrading/assembly-downgrade.adoc:17]: I've simplified to:

```
You can perform a downgrade to:

. Revert your Cluster Operator to the previous Strimzi version.
** xref:proc-downgrade-cluster-operator-{context}[]

. Downgrade all Kafka brokers and client applications to the previous Kafka version.
** xref:assembly-downgrade-kafka-versions-{context}[]
```

Comment (PaulRMellor) [reply documentation/assemblies/upgrading/assembly-downgrade.adoc:17]: This section won't be in the doc yet

File: documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc (hunk 12 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc:57]: Removed from this procedure and added a new generic procedure to the CO downgrade procedure.

image

File: documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc (hunk 13 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:8]: Changed to:

```
If you are encountering issues with Strimzi,
you can revert your installation.
```

File: documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc (hunk 14 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:73]: Yes. Once it's cherry-picked, we can bring the kafka downgrade sections back in and improve the chapter a whole.

**Review by scholzj — APPROVED** [review]

File: documentation/deploying/master.adoc [hunk 15]
```diff
@@ -20,3 +20,5 @@ include::assemblies/deploying/assembly-deploy-verify.adoc[leveloffset=+1]
 include::assemblies/metrics/assembly-metrics.adoc[leveloffset=+1]
 //Upgrading the deployment
 include::assemblies/upgrading/assembly-upgrade.adoc[leveloffset=+1]
+//Downgrading the deployment
+include::modules/upgrading/proc-downgrade-cluster-operator.adoc[leveloffset=+1]
```
Comment (laidan6000) [inline documentation/deploying/master.adoc:24]: Just checking that the correct module is included in the `master.adoc`?

File: documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+. Revert your custom resources to reflect the supported configuration options available for the version of Strimzi you are downgrading to.
+
+. Update the Cluster Operator.
+
+.. Modify the installation files for the previous version according to the namespace the Cluster Operator is running in.
++
+include::snip-cluster-operator-namespace-sed.adoc[]
++
+.. If you modified one or more environment variables in your existing Cluster Operator `Deployment`, edit the
+`install/cluster-operator/060-Deployment-cluster-operator.yaml` file to use those environment variables.
```
Comment (laidan6000) [inline documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:32]: In Strimzi 0.19.0, this file is named `050-Deployment-strimzi-cluster-operator.yaml`.

File: documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+. Get the image for the Kafka pod to ensure the downgrade was successful:
++
+[source,shell,subs="+quotes,attributes"]
+----
+kubectl get po my-cluster-kafka-0 -o jsonpath='{.spec.containers[0].image}'
+----
++
+The image tag shows the new Strimzi version followed by the Kafka version. For example, `_NEW-STRIMZI-VERSION_-kafka-_CURRENT-KAFKA-VERSION_`.
+
+You now have an updated Cluster Operator.
```
Comment (laidan6000) [inline documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:52]: _Suggested change:_

Your Cluster Operator was downgraded to the previous version.

File: documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc [hunk 18]
```diff
@@ -67,6 +68,9 @@ spec:
 ----
 <1> This is changed to the new version
 <2> This remains at the current version
+--
++
+WARNING: You cannot downgrade Kafka once `spec.kafka.version` has changed, as the `inter.broker.protocol.version` for the new Kafka version will differ. The protocol version determines the schemas used for persistent metadata stored by the broker, including messages written to `__consumer_offsets`. The cluster will not understand the messages.
```
Comment (laidan6000) [inline documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:73]:
```suggestion
WARNING: You cannot downgrade Kafka after `spec.kafka.version` has changed, because the `inter.broker.protocol.version` for the new Kafka version will differ. The inter-broker protocol version determines the schemas used for persistent metadata stored by the broker, including messages written to `__consumer_offsets`. The downgraded cluster will not understand the messages.
```

File: documentation/modules/upgrading/con-downgrade-target-version.adoc [hunk 19]
```diff
@@ -4,15 +4,40 @@
 
 [id='com-target-downgrade-version-{context}']
```
Comment (laidan6000) [inline documentation/modules/upgrading/con-downgrade-target-version.adoc:5]: Typo in module ID: con.

File: documentation/modules/upgrading/con-downgrade-target-version.adoc [hunk 20]
```diff
@@ -4,15 +4,40 @@
...
+and the state at which messages have been logged.
 
-* If the target downgrade version of Kafka has the same `log.message.format.version` as the current version, the Cluster Operator downgrades by performing a single rolling restart of the brokers.
-* If the target downgrade version of Kafka has a different `log.message.format.version`, downgrading is only possible if the running cluster has _always_ had `log.message.format.version` set to the version used by the downgraded version.
-+
+You cannot revert to the previous Kafka version if the `inter.broker.protocol.version` differs, or messages have been added
+to message logs that use a newer `log.message.format.version`.
+
+The `inter.broker.protocol.version` determines the schemas used for persistent metadata stored by the broker, such as the schema for messages written to `__consumer_offsets`.
+If you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the cluster will not work as it will not understand the messages.
```
Comment (laidan6000) [inline documentation/modules/upgrading/con-downgrade-target-version.adoc:16]:
```suggestion
If you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the cluster will not work because it will not understand the messages.
```

**Review by laidan6000 — CHANGES_REQUESTED** [review]
Much clearer overall. A few questions sprung to mind on a second reading.

File: documentation/deploying/master.adoc (hunk 15 above)
Comment (PaulRMellor) [reply documentation/deploying/master.adoc:24]: For the time being, yes. There's a new module for downgrading, but we're concentrating only on the CO downgrade for the time being. Other files exists, but won't be included yet.

File: documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc (hunk 16 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:32]: Still 060, but new name.

**Review by laidan6000 — APPROVED** [review]
Changes look good.

**Comment by PaulRMellor** [comment]
Kafka downgrade sections reintroduced, following the work to make specifying `inter.broker.protocol.version` explicit in https://github.com/strimzi/strimzi-kafka-operator/pull/4044 and https://github.com/strimzi/strimzi-kafka-operator/pull/4058

File: documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc [hunk 21]
```diff
@@ -42,14 +44,12 @@ Wait for the rolling updates to complete.
 kubectl get po my-cluster-kafka-0 -o jsonpath='{.spec.containers[0].image}'
```
Comment (PaulRMellor) [inline documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:44]: @tombentley - we do the check on the Kafka pod for the CO upgrade?

File: documentation/modules/upgrading/con-downgrade-target-version.adoc [hunk 22]
```diff
@@ -2,17 +2,42 @@
...
-+
+
 ** Two rolling restarts of the brokers if the interbroker protocol of the two versions is different
 ** A single rolling restart if they are the same
+
+Downgrading is _not possible_ if the new version has ever used a `log.message.format.version` that is not supported by the previous version, including when the default value for `log.message.format.version` is used. For example, this resource can be downgraded to Kafka version {KafkaVersionLower} because the `log.message.format.version` has not been changed:
+
+[source,yaml,subs=attributes+]
+----
+apiVersion: v1alpha1
```
Comment (PaulRMellor) [inline documentation/modules/upgrading/con-downgrade-target-version.adoc:32]: @scholzj api version check on the upgrade and downgrade procedures. We have `v1alpha1` in our examples

Comment (scholzj) [reply documentation/modules/upgrading/con-downgrade-target-version.adoc:32]: Aren't we normally using some variable? It should definitely be `v1beta1`.

File: documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc (hunk 21 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:44]: It was something that @michalxo mentioned when reviewing the procedure. He wondered why we don't do the verification on the CO pod. 

```
kubectl get pod <cluster-operator-pod-name> -o jsonpath='{.spec.containers[0].image}'
        quay.io/strimzi/operator:latest%
```

File: documentation/modules/upgrading/con-downgrade-target-version.adoc (hunk 22 above)
Comment (PaulRMellor) [reply documentation/modules/upgrading/con-downgrade-target-version.adoc:32]: I've replaced with the {KafkaApiVersion} variable.

**Comment by scholzj** [comment]
@tombentley Does this look good now? Were your comments addressed?

File: documentation/modules/upgrading/con-downgrade-target-version.adoc [hunk 23]
```diff
@@ -2,17 +2,42 @@
...
+and the state at which messages have been logged.
 
-* If the target downgrade version of Kafka has the same `log.message.format.version` as the current version, the Cluster Operator downgrades by performing a single rolling restart of the brokers.
-* If the target downgrade version of Kafka has a different `log.message.format.version`, downgrading is only possible if the running cluster has _always_ had `log.message.format.version` set to the version used by the downgraded version.
-+
+You cannot revert to the previous Kafka version if the `inter.broker.protocol.version` differs, or messages have been added
+to message logs that use a newer `log.message.format.version`.
+
+The `inter.broker.protocol.version` determines the schemas used for persistent metadata stored by the broker, such as the schema for messages written to `__consumer_offsets`.
+If you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the cluster will not work because it will not understand the messages.
```
Comment (tombentley) [inline documentation/modules/upgrading/con-downgrade-target-version.adoc:16]: An "earlier IBP" isn't quite the right language. Each version of Kafka supports a range of IBPs, so "earlier" is a little ambiguous (it could be taken to mean earlier in the range of IBPs supported by a particular version). 

```suggestion
If you downgrade to a version of Kafka that doesn't understand an `inter.broker.protocol.version` that has (ever) been previously used in the cluster the broker will encounter data they cannot understand.
```

File: documentation/modules/upgrading/con-downgrade-target-version.adoc [hunk 24]
```diff
@@ -2,17 +2,42 @@
...
+= Kafka version compatibility for downgrades
 
-How the Cluster Operator handles a downgrade operation depends on the `log.message.format.version`.
+Kafka downgrades are dependent on compatible current and target xref:ref-kafka-versions-{context}[Kafka versions],
+and the state at which messages have been logged.
 
-* If the target downgrade version of Kafka has the same `log.message.format.version` as the current version, the Cluster Operator downgrades by performing a single rolling restart of the brokers.
-* If the target downgrade version of Kafka has a different `log.message.format.version`, downgrading is only possible if the running cluster has _always_ had `log.message.format.version` set to the version used by the downgraded version.
-+
+You cannot revert to the previous Kafka version if the `inter.broker.protocol.version` differs, or messages have been added
```
Comment (tombentley) [inline documentation/modules/upgrading/con-downgrade-target-version.adoc:12]: This sentence on its own (I know is clarified in a couple of sentences time) might make the reader think they can just change the IBP first. 

```suggestion
You cannot revert to the previous Kafka version if that version don't support any of the `inter.broker.protocol.version` settings which have ever been used in that cluster, or messages have been added
```

File: documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc (hunk 21 above)
Comment (tombentley) [reply documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:44]: Still not understanding this. `get po` and `get pod` are the same thing. Or are you referring to the `quay.io/strimzi/operator:latest%` bit?

Comment (michalxo) [reply documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:44]: Looking at my comment in Jira, I think the latter is what I meant to suggest for a change ->
``` wrong pod used in command?. As we upgraded strimzi CO and not kafka (that did not change)

$ kubectl get pod <cluster-operator-pod-name> -o jsonpath='{.spec.containers[0].image}'
        quay.io/strimzi/operator:latest%
```

Comment (scholzj) [reply documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:44]: IMHO the upgrade of the operator is successful after all operand pods roll to the new images. So checking the image in the Kafka pods makes more sense than checking it in the Cluster Operator pod.

Comment (tombentley) [reply documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:44]: Ah, finally I understand what's being asked here! (Sorry if I was being dense). Yes, checking on the operand pods makes more sense.

Comment (scholzj) [reply documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:44]: @PaulRMellor If you do any edits to this, I think using the full name - `kubectl get pod` makes more sense in documentation.

Comment (PaulRMellor) [reply documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:44]: @tombentley @scholzj @michalxo  -- Sorry for the confusion and thanks for the comments. I've left the commands as is, but changed to `pod` rather than `po`

**Comment by PaulRMellor** [comment]
@tombentley - do you have any more comments?

File: documentation/modules/upgrading/con-downgrade-target-version.adoc [hunk 25]
```diff
@@ -2,17 +2,42 @@
...
+= Kafka version compatibility for downgrades
 
-How the Cluster Operator handles a downgrade operation depends on the `log.message.format.version`.
+Kafka downgrades are dependent on compatible current and target xref:ref-kafka-versions-{context}[Kafka versions],
+and the state at which messages have been logged.
 
-* If the target downgrade version of Kafka has the same `log.message.format.version` as the current version, the Cluster Operator downgrades by performing a single rolling restart of the brokers.
-* If the target downgrade version of Kafka has a different `log.message.format.version`, downgrading is only possible if the running cluster has _always_ had `log.message.format.version` set to the version used by the downgraded version.
-+
+You cannot revert to the previous Kafka version if that version don't support any of the `inter.broker.protocol.version` settings which have ever been used in that cluster,
```
Comment (tombentley) [inline documentation/modules/upgrading/con-downgrade-target-version.adoc:12]:
```suggestion
You cannot revert to the previous Kafka version if that version don't support any of the `inter.broker.protocol.version` settings which have _ever been used_ in that cluster,
```

**Review by tombentley — APPROVED** [review]

**Comment by scholzj** [comment]
@PaulRMellor Looks like the CI is failing:

```
./.azure/scripts/check_docs.sh
Avoid 'nt contraction:
documentation/modules/upgrading/con-downgrade-target-version.adoc:16:If you downgrade to a version of Kafka that doesn't understand an `inter.broker.protocol.version` that has (ever) been previously used in the cluster the broker will encounter data they cannot understand.
ERROR: 1 docs problems found.
Makefile:138: recipe for target 'docu_check' failed
```
